### PR TITLE
Update name of tokens on token list

### PR DESCRIPTION
**Pull request by tommasini**
**PR Title**

- A brief description of changes. If the PR has breaking changes add `BREAKING:`
  to the start of the PR title.

**Description**
This PR aims to update the token name of the tokens on the token list.

- BREAKING:
It was added the name property on the follow function:
* `addToken` - now it receives an object with optional params (name, image)

- CHANING:
* `addTokens`

It was created a new function called
`updateTokensAttribute` - It aims to update the tokens attribute to the tokens that are on the state with the fetched token list.

It was created a new ERC20 standard to get the token name
`getTokenName`
and the function to call it on Assets Controller
`getERC20TokenName`

**Checklist**

- [x] Tests are included if applicable
- [x] Any added code is fully documented

**Issue**

Resolves #???

### CODE REVIEW

File: packages/assets-controllers/src/TokensController.test.ts [hunk 1]
```diff
@@ -1222,4 +1256,100 @@ describe('TokensController', () => {
...
+        name: '',
+        symbol: 'GHI',
+        aggregators: [],
+        decimals: 0,
+        iconUrl: '',
+        occurrences: 1,
+      },
+    ];
+
+    it('update token name when a matching token is found in the tokenList', async () => {
```
Comment (mcmire) [inline packages/assets-controllers/src/TokensController.test.ts:1290]: By using the word "when" in this test, you seem to be referencing some kind of logic. So, what happens when a matching token is _not_ found in the tokenList? In other words, do you need a test that exercises the opposite case?

File: packages/assets-controllers/src/TokensController.ts [hunk 2]
```diff
@@ -344,8 +347,7 @@ export class TokensController extends BaseController<
    *
    * @param tokensToImport - Array of tokens to import.
    */
-  async addTokens(tokensToImport: Token[]) {
-    const releaseLock = await this.mutex.acquire();
```
Comment (mcmire) [inline packages/assets-controllers/src/TokensController.ts:348]: Looks like we are removing the mutex lock here. Is that intentional? If so, what's the rationale?

File: packages/assets-controllers/src/TokensController.ts [hunk 3]
```diff
@@ -457,8 +457,15 @@ export class TokensController extends BaseController<
 
     try {
       incomingDetectedTokens.forEach((tokenToAdd) => {
-        const { address, symbol, decimals, image, aggregators, isERC721 } =
-          tokenToAdd;
+        const {
```
Comment (mcmire) [inline packages/assets-controllers/src/TokensController.ts:460]: Can we add a test for this change? Apologies if it's being tested implicitly — it's not clear to me.

File: packages/assets-controllers/src/TokensController.ts [hunk 4]
```diff
@@ -264,13 +264,15 @@ export class TokensController extends BaseController<
...
+   * @param name - Name of the token.
    * @param image - Image of the token.
    * @returns Current token list.
    */
   async addToken(
     address: string,
     symbol: string,
     decimals: number,
+    name?: string,
     image?: string,
```
Comment (mcmire) [inline packages/assets-controllers/src/TokensController.ts:276]: Instead of adding another argument smack-dab in the middle of the argument list, for backward compatibility, what do you think about adding it at the end? That way it can be left off and nothing bad will happen:

``` suggestion
    image?: string,
    name?: string,
```

Alternatively, if we don't mind introducing a backward incompatible change, what do you feel about putting `image` and `name` into an object? Generally speaking options are easier to work with inside of an object than as positional arguments. Otherwise, someone who wants to _just_ pass `image` but not `name` would have to pass `undefined`, which feels kind of ugly.

In other words perhaps we want to say:

``` suggestion
    { name, image }: { name?: string: image?: string } = {}
```

File: packages/assets-controllers/src/TokensController.ts (hunk 2 above)
Comment (tommasini) [reply packages/assets-controllers/src/TokensController.ts:348]: No, not intentional sorry! It missed me somehow

**Comment by tommasini** [comment]
Thanks for the review @mcmire I will address your comments later today

File: packages/assets-controllers/src/TokensController.ts (hunk 4 above)
Comment (tommasini) [reply packages/assets-controllers/src/TokensController.ts]: Amazing suggestion! Did the refactor!

File: packages/assets-controllers/src/TokensController.test.ts (hunk 1 above)
Comment (tommasini) [reply packages/assets-controllers/src/TokensController.test.ts:1301]: Added an opposite test case! Amazing catch!

File: packages/assets-controllers/src/TokensController.ts (hunk 2 above)
Comment (tommasini) [reply packages/assets-controllers/src/TokensController.ts]: Added again the mutex logic

File: packages/assets-controllers/src/TokensController.ts (hunk 3 above)
Comment (tommasini) [reply packages/assets-controllers/src/TokensController.ts:464]: Updated the existing test of 'addTokens', you were right with was not being test properly!

File: packages/assets-controllers/src/TokensController.ts [hunk 5]
```diff
@@ -546,6 +558,24 @@ export class TokensController extends BaseController<
     return tokens[tokenIndex];
   }
 
+  /**
+   * This is a function that updates the tokens name for the tokens name if it is not defined.
+   *
+   * @param tokenList - Represents the fetched token list from service API
+   */
+  updateTokensName(tokenList: TokenListMap) {
```
Comment (Gudahtt) [inline packages/assets-controllers/src/TokensController.ts:566]: Just a suggestion, but, have you considered adding an `updateTokens` method instead? This would let us update token metadata including the name.

It just seemed a bit strange from an API design standpoint to have a function for updating a single property on each token, yet not any other properties.

File: packages/assets-controllers/src/TokensController.ts [hunk 6]
```diff
@@ -664,7 +694,12 @@ export class TokensController extends BaseController<
       switch (suggestedAssetMeta.type) {
         case 'ERC20':
           const { address, symbol, decimals, image } = suggestedAssetMeta.asset;
-          await this.addToken(address, symbol, decimals, image);
+          /**
+           *
+           * The reason for this undefined is that the wallet_watchAsset doesn't return the name
```
Comment (Gudahtt) [inline packages/assets-controllers/src/TokensController.ts:699]: Nit: suggested rewording to highlight that this comment is referring to `name`, which was a little bit unclear to me at first
```suggestion
           * The reason for `name` being omitted is that `wallet_watchAsset` doesn't return the name
```

File: packages/assets-controllers/src/TokensController.ts (hunk 5 above)
Comment (tommasini) [reply packages/assets-controllers/src/TokensController.ts:566]: Thanks for the review @Gudahtt , I didn't considered that, something like this? 
` updateTokensAtribute(
    tokenList: TokenListMap,
    tokenAttribute: keyof Token & keyof TokenListToken,
  ) {
    const { tokens } = this.state;

    const newTokens = tokens.map((token) => {
      const newToken = tokenList[token.address.toLowerCase()];

      return !token[tokenAttribute] && newToken?.[tokenAttribute]
        ? { ...token, [tokenAttribute]: newToken[tokenAttribute] }
        : { ...token };
    });

    this.update({ tokens: newTokens });
  }
`

It's seems way better, let me know if you aggree

Comment (Gudahtt) [reply packages/assets-controllers/src/TokensController.ts:566]: Thanks @tommasini , that would make sense to me as well.

Though upon looking at this further, I'm having trouble grasping why the token detection controller is playing the role of updating token metadata here. I would expect the token detection controller to _just_ handle detection, not to synchronize metadata between the token list controller and the token controller. I see that this is a strange case though, sort of like a migration, since this `name` field was added after the fact. Hmm. Maybe I'll start a thread on Slack to discuss this further.

Comment (tommasini) [reply packages/assets-controllers/src/TokensController.ts:566]: Yes please, I'm indeed trying to cover the tokens that already existed and did not have the name property, since the UI of the [token list changed](https://recordit.co/H30qk9FcyH) (at least on mobile)

**Comment by mcmire** [comment]
@tommasini Cool, just post another comment when you're ready for review again!

**Comment by tommasini** [comment]
@mcmire @Gudahtt sorry for taking so long on this PR, when you are able please can you take a look on it?

**Review by bergarces — COMMENTED** [review]
No red flags from me. It should not negatively impact `wallet_watchAsset`.
